### PR TITLE
Adding content length validation

### PR DESCRIPTION
**Pull request by alyssawilk**
Changing H2 from tearing down the connection on invalid content length (and other HTTP style errors) to sending a reset stream instead.
Rejecting POSTs and PUTs with no body and no chunked encoding with 411 (Length Required)
Adding tests for invalid, duplicate, and missing content lengths.

### CODE REVIEW

File: source/common/http/http1/codec_impl.cc [hunk 1]
```diff
@@ -411,6 +411,11 @@ int ServerConnectionImpl::onHeadersComplete(HeaderMapImplPtr&& headers) {
         http_parser_pause(&parser_, 1);
       }
 
+    } else if (headers->Method()->value() == "POST" || headers->Method()->value() == "PUT") {
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: To avoid the string compares here we can pull the already parsed method out of the http_parser struct.
